ebr_trackerbot/storage/db.py

```diff
@@ -0,0 +1,135 @@
+"""
+Slack Bot Generic DB Storage
+"""
+
+import logging
+import pendulum
+import sys
+
+
+def create_table(get_connection):
+    """
+    Create table if it does not exist
+    """
+    conn = get_connection()
+    try:
+        cursor = conn.cursor()
+        cursor.execute(
+            "CREATE TABLE IF NOT EXISTS tracks "
+            + "(id integer PRIMARY KEY, user text NOT NULL, test text NOT NULL, "
+            + " expiry timestamp NOT NULL, channel_id text NOT NULL, thread_ts text NOT NULL);"
+        )
+        conn.commit()
+    except:
+        logging.error(str(sys.exc_info()[1]))
+        conn.close()
+
+
+def delete_for_user(get_connection, user, test):
+    """
+    Delete tracking for user and test
+    """
+    logging.debug("Delete track for user: " + user + ", test: " + test)
+    conn = get_connection()
+    try:
+        cursor = conn.cursor()
+        cursor.execute(""" DELETE FROM tracks WHERE user = ? AND test = ?; """, [user, test])
+        conn.commit()
+    except:
+        logging.error(sys.exc_info()[1])
+        conn.close()
+
+
+def save(get_connection, user, data):
+    """
+    Creates tracking for user and test
+    """
+    logging.debug("Saving track for user: %s, data: ", user)
+    logging.debug(data)
+    conn = get_connection()
+    try:
+        cursor = conn.cursor()
+        cursor.execute(
+            """ SELECT expiry, channel_id, thread_ts FROM tracks WHERE user = ? AND test = ? """, [user, data["test"]]
+        )
+        rows = cursor.fetchall()
+        if not rows:
+            cursor.execute(
+                """ INSERT INTO tracks (user, test, expiry, channel_id, thread_ts) VALUES (?, ?, ?, ?, ?); """,
+                [user, data["test"], data["expiry"], data["channel_id"], data["thread_ts"]],
+            )
+        else:
+            cursor.execute(
+                """ UPDATE tracks SET expiry = ?, channel_id = ?, thread_ts = ? WHERE user = ? AND test = ? """,
+                [data["expiry"], data["channel_id"], data["thread_ts"], user, data["test"]],
+            )
+        conn.commit()
+    except:
+        logging.error(sys.exc_info()[1])
+        conn.close()
+
+
+def load_all_tracked_tests(get_connection):
+    """
+    Load all tracked tests
+    """
+    result = []
+    conn = get_connection()
+    try:
+        cursor = conn.cursor()
+        cursor.execute(""" SELECT user, test, expiry, channel_id, thread_ts FROM tracks; """)
+        rows = cursor.fetchall()
+        for row in rows:
+            data = {
+                "user": row[0],
```

Comment:
I think the cursor returns a named tuple so you can use here user, test, etc. instead indices.
Makes your code safe against accidental order changes in the select

Reply:
I double checked it and fetchall returns only list no named tuple :(

Reply:
Ok, it is indeed no named tuple but a list. The docs say something different: https://github.com/mkleehammer/pyodbc/wiki/Cursor#fetchall

And based on the code it says you can use column names to access the column values: https://github.com/mkleehammer/pyodbc/blob/master/src/cursor.h#L152